checks/shell_download_validate.go

```diff
@@ -0,0 +1,628 @@
+}
+
+func extractCommand(cmd interface{}) ([]string, bool) {
+	c, ok := cmd.(*syntax.CallExpr)
+	if !ok {
+		return nil, ok
+	}
+
+	var ret []string
+	for _, w := range c.Args {
+		if len(w.Parts) != 1 {
+			continue
+		}
+		lit, ok := w.Parts[0].(*syntax.Lit)
+		if ok && !strings.EqualFold(lit.Value, "sudo") {
+			ret = append(ret, lit.Value)
+		}
+	}
+	return ret, true
+}
+
+func isFetchPipeExecute(node syntax.Node, cmd, pathfn string,
+	logf func(s string, f ...interface{})) bool {
+	// BinaryCmd {Op=|, X=CallExpr{Args={curl, -s, url}}, Y=CallExpr{Args={bash,}}}.
+	bc, ok := node.(*syntax.BinaryCmd)
+	if !ok {
+		return false
+	}
+
+	// Look for the pipe operator.
+	if !strings.EqualFold(bc.Op.String(), "|") {
+		return false
+	}
+
+	leftStmt, ok := extractCommand(bc.X.Cmd)
+	if !ok {
+		return false
+	}
+	rightStmt, ok := extractCommand(bc.Y.Cmd)
+	if !ok {
+		return false
+	}
+
+	if !isDownloadUtility(leftStmt) {
+		return false
+	}
+
+	if !isInterpreter(rightStmt) {
+		return false
+	}
+
+	logf("!! frozen-deps/fetch-execute - %v is fetching and executing non-pinned program '%v'",
+		pathfn, cmd)
+	return true
+}
+
+func getRedirectFile(red []*syntax.Redirect) (string, bool) {
+	if len(red) == 0 {
+		return "", false
+	}
+	for _, r := range red {
+		if !strings.EqualFold(r.Op.String(), ">") {
+			continue
+		}
+
+		if len(r.Word.Parts) != 1 {
+			continue
+		}
+
+		lit, ok := r.Word.Parts[0].(*syntax.Lit)
+		if ok {
+			return lit.Value, true
+		}
+	}
+	return "", false
+}
+
+func isExecuteFiles(node syntax.Node, cmd, pathfn string, files map[string]bool,
+	logf func(s string, f ...interface{})) bool {
+	ce, ok := node.(*syntax.CallExpr)
+	if !ok {
+		return false
+	}
+
+	c, ok := extractCommand(ce)
+	if !ok {
+		return false
+	}
+
+	ok = false
+	for fn := range files {
+		if isInterpreterWithFile(c, fn) || isExecuteFile(c, fn) {
+			logf("!! frozen-deps/fetch-execute - %v is fetching and executing non-pinned program '%v'",
+				pathfn, cmd)
+			ok = true
+		}
+	}
+
+	return ok
+}
+
+// Detect `fetch | exec`.
+func validateCommandIsNotFetchPipeExecute(cmd, pathfn string, logf func(s string, f ...interface{})) (bool, error) {
+	in := strings.NewReader(cmd)
+	f, err := syntax.NewParser().Parse(in, "")
+	if err != nil {
+		return false, ErrParsingShellCommand
+	}
+
+	cmdValidated := true
+	syntax.Walk(f, func(node syntax.Node) bool {
+		if isFetchPipeExecute(node, cmd, pathfn, logf) {
+			cmdValidated = false
+		}
+
+		// Continue walking the node graph.
+		return true
+	})
+
+	return cmdValidated, nil
+}
+
+func recordFetchFileFromNode(node syntax.Node) (pathfn string, ok bool, err error) {
+	ss, ok := node.(*syntax.Stmt)
+	if !ok {
+		return "", false, nil
+	}
+
+	cmd, ok := extractCommand(ss.Cmd)
+	if !ok {
+		return "", false, nil
+	}
+
+	if !isDownloadUtility(cmd) {
+		return "", false, nil
+	}
+
+	fn, ok := getRedirectFile(ss.Redirs)
+	if !ok {
+		return getOutputFile(cmd)
+	}
+
+	return fn, true, nil
```

Comment:
do we need to the boolean at all? can callers just check that `fn != ""` ?

Reply:
the caller could do that, yes. I was not sure what's the `golang` way of doing things. Having a separate `bool` means we're not conflating the output string with the `is found` information. But you're right in this function, the empty string indicates it was not found. Let me know if I should change it.

Reply:
so I think it may be better to keep the current version, as my concern is the same as https://github.com/golang/go/wiki/CodeReviewComments#in-band-errors. wdut?

Reply:
I think this still feels unnecessarily complicated, because there is both a boolean and an error. 

are there ambiguities here with empty value being a possibly valid return value? If not, probably cleaner to remove. I think that is the point that https://github.com/golang/go/wiki/CodeReviewComments#in-band-errors is making.